encode.go

```diff
@@ -92,34 +110,33 @@ func Encode(dst, src []byte) []byte {
 		if len(src) != 0 {
 			d += emitLiteral(dst[d:], src)
 		}
+		e.cur += 4
 		return dst[:d]
 	}
 
-	// Initialize the hash table. Its size ranges from 1<<8 to 1<<14 inclusive.
-	const maxTableSize = 1 << 14
-	shift, tableSize := uint(32-8), 1<<8
-	for tableSize < maxTableSize && tableSize < len(src) {
-		shift--
-		tableSize *= 2
+	// Ensure that e.cur doesn't wrap.
+	if e.cur > 1<<30 {
```

Comment:
It's still not obvious to me what the invariant is before/after every call to encoder.encode. In other words, what is the relationship between e.cur and the elements of e.table?

Originally, I thought that e.table[i] <= e.cur for all i in the range [0, tableSize), but that's not true, because you can reset e.cur to zero without touching e.table's elements. Therefore, I'm not entirely convinced that any reference to the previous block is discarded by the t < 0 check.

Whatever the relationship is, I think it needs better commenting (probably as doc comments near "type encoder"'s fields), and almost surely some explicit testing, especially around the wrap point. Otherwise, some future maintainer, whether me, you or somebody else, might make a change that violates whatever the unwritten assumptions are, and introduce a subtle bug.